planner/core/logical_plan_builder.go

```diff
@@ -2421,6 +2421,95 @@ func (b *PlanBuilder) resolveCorrelatedAggregates(ctx context.Context, sel *ast.
 	return correlatedAggMap, nil
 }
 
+type asofResolver struct {
+	ctx      sessionctx.Context
+	tsValues []*types.Datum
+	err      error
+}
+
+func (a *asofResolver) Enter(inNode ast.Node) (ast.Node, bool) {
+	switch n := inNode.(type) {
+	case *ast.AsOfClause:
+		if n.TsExpr != nil {
+			tsRes, err := evalAstExpr(a.ctx, n.TsExpr)
+			if err != nil {
+				a.err = err
+				return n, true
+			}
+			a.tsValues = append(a.tsValues, &tsRes)
+
+		}
+		return n, false
+	// for the cases: select * from a as of timestamp 'xxx', b;
+	case *ast.TableName:
+		if n.AsOf == nil {
+			a.tsValues = append(a.tsValues, nil)
```

Comment:
I think we could add simple `hasCurrentTS` boolean flag in the resolver. Then the condition is like `hasCurrentTS == true` or `len(tsValues) > 0`, but not both.